### PR TITLE
[`GPT-J`] Fix causal mask dtype

### LINKED ISSUE #23136: [GPT-J] where expected condition to be a boolean tensor, but got a tensor with dtype Half
### System Info

- `transformers` version: 4.28.1
- Platform: Linux-5.10.147+-x86_64-with-glibc2.31
- Python version: 3.10.11
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.0+cu118 (True)
- Tensorflow version (GPU?): 2.12.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.6.9 (gpu)
- Jax version: 0.4.8
- JaxLib version: 0.4.7
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

@younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Collab example - [Link](https://colab.research.google.com/drive/1D689vHxZk5Bov5piIXOQ62t3M_cim4dr?usp=sharing)

### Expected behavior

I expected some output generated from the model

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/23136

When going for `low_cpu_mem_usage` each parameter is force-casted to the expected dtype, which is force-set to `torch.float16` for 8bit models.

Therefore, for 8bit models (and also half-precision models) the causal mask is always force casted to float16 as it is part of the model's state dict, hence expected to be loaded from the Hub if the mask is available on the state dict.

The fix is to add `persistant=False` and add a field `_keys_to_ignore_on_unexpected` (for removing the warnings) to avoid loading that causal mask from the state dict and assign it to the buffer, and all causal masks that are saved as buffers should do the same to avoid unexpected behaviors.

cc @sgugger

### CODE REVIEW

File: src/transformers/models/gptj/modeling_gptj.py [hunk 1]
```diff
@@ -733,6 +734,7 @@ def custom_forward(*inputs):
 )
 class GPTJForCausalLM(GPTJPreTrainedModel):
     _keys_to_ignore_on_load_missing = [r"h\.\d+\.attn\.masked_bias", r"h\.\d+\.attn\.bias"]
```
Comment (sgugger) [inline src/transformers/models/gptj/modeling_gptj.py:736]: I think you can remove this now, since they won't be missing anymore.

**Review by sgugger — APPROVED** [review]
Thanks for the fix!
